dbos/dbos.go

```diff
@@ -372,62 +373,90 @@ func (c *dbosContext) Launch() error {
 	return nil
 }
 
-// Cancel gracefully shuts down the DBOS runtime by canceling the context, waiting for
-// all workflows to complete, and cleaning up system resources including the database
-// connection pool, queue runner, workflow scheduler, and admin server.
+// Cancel cancels the context, waits for workflows to complete, and stops the queue runner and workflow scheduler.
+// This is the core cancellation functionality that signals workflows should stop processing.
 // All workflows and steps contexts will be canceled, which one can check using their context's Done() method.
-//
-// This method blocks until all workflows finish and all resources are properly cleaned up.
-// It should be called when the application is shutting down to ensure data consistency.
-func (c *dbosContext) Cancel() {
-	c.logger.Info("Shutting down DBOS context")
+// Cancel is a permanent operation and should be called when the application is shutting down or deactivated.
+func (c *dbosContext) Cancel(timeout time.Duration) {
+	c.logger.Info("Cancelling DBOS context")
 
 	// Cancel the context to signal all resources to stop
-	c.ctxCancelFunc(errors.New("DBOS shutdown initiated"))
+	c.ctxCancelFunc(errors.New("DBOS cancellation initiated"))
 
 	// Wait for all workflows to finish
 	c.logger.Info("Waiting for all workflows to finish")
-	c.workflowsWg.Wait()
-	c.logger.Info("All workflows completed")
+	done := make(chan struct{})
+	go func() {
+		c.workflowsWg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+		c.logger.Info("All workflows completed")
+	case <-time.After(timeout):
+		// For now just log a warning: eventually we might want Cancel to return an error.
+		c.logger.Warn("Timeout waiting for workflows to complete", "timeout", timeout)
+	}
+
+	// Wait for queue runner to finish
+	if c.queueRunner != nil {
+		c.logger.Info("Waiting for queue runner to complete")
+		select {
+		case <-c.queueRunner.completionChan:
+			c.logger.Info("Queue runner completed")
+			c.queueRunner = nil
+		case <-time.After(timeout):
+			c.logger.Warn("Timeout waiting for queue runner to complete", "timeout", timeout)
```

Comment:
we cannot safely set the pointer to `nil` in case of timeout: the resource might yet to be accessed, which would cause a panic.